src/qubes_fwupdmgr.py

```diff
@@ -0,0 +1,502 @@
+                    raise ValueError()
+            except ValueError:
+                print("Invalid choice.")
+
+        if downgrade:
+            while True:
+                try:
+                    releases = updates_list[device_number]["Releases"]
+                    for i, fw_dngd in enumerate(releases):
+                        print(decorator)
+                        print(
+                            "  %s. Firmware downgrade version:\t %s" %
+                            (i+1, fw_dngd["Version"])
+                        )
+                        description = fw_dngd["Description"].replace("<p>", "")
+                        description = description.replace("<li>", "")
+                        description = description.replace("<ul>", "")
+                        description = description.replace("</ul>", "")
+                        description = description.replace("</p>", "\n   ")
+                        description = description.replace("</li>", "\n   ")
+                        print("   Description: %s" % description)
+                    print("If you want to abandon downgrade process press N.")
+                    choice = input(
+                        "Otherwise choose downgrade number: "
+                    )
+                    if choice == 'N' or choice == 'n':
+                        return 99
+                    downgrade_number = int(choice)-1
+                    if 0 <= downgrade_number < len(releases):
+                        return device_number, downgrade_number
+                    else:
+                        raise ValueError()
+                except ValueError:
+                    print("Invalid choice.")
+
+    def _parse_parameters(self, updates_list, choice):
+        """Parses device name, url, version and SHA1 checksum of the file list.
+
+        Keywords arguments:
+        updates_list - list of updates for specified device
+        choice -- number of device to be updated
+        """
+        self.name = updates_list[choice]["Name"]
+        self.version = updates_list[choice]["Releases"][0]["Version"]
+        for ver_check in updates_list[choice]["Releases"]:
+            if l_ver(ver_check["Version"]) >= l_ver(self.version):
+                self.version = ver_check["Version"]
+                self.url = ver_check["Url"]
+                self.sha = ver_check["Checksum"]
+
+    def _install_firmware_update(self, path):
+        """Installs firmware update for specified device.
+
+        Keywords arguments:
+        path - absolute path to firmware update archive
+        """
+        cmd_install = [
+            "/bin/fwupdmgr",
+            "install",
+            path
+        ]
+        p = subprocess.Popen(cmd_install)
+        p.wait()
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Firmware update failed")
+
+    def _read_dmi(self):
+        """Reads BIOS information from DMI."""
+        cmd_dmidecode = [
+            "dmidecode",
```

Comment:
Have you checked for python module?

Reply:
There is external wrapper for dmidecode. It uses similar [subprocess](https://github.com/zaibon/py-dmidecode/blob/master/dmidecode.py#L210)